### PR TITLE
Factor out a lot of stuff to separate files

### PR DESCRIPTION
**Pull request by Jalle19**
This is a step in completing the refactoring roadmap we laid out a while ago. Here's a summary of the changes:

* channels, tags, recordings, timers, events and schedules have been factored out into classes living under the `tvheadend::entity` namespace. The naming here is up for debate. Basically an entity is something that can be marked as dirty, so all of these classes extend the base `Entity` class which provides that functionality.

* the new `Schedule` class has a new concept, `Segment`. A segment is a part of the schedule. The class provides a `GetSegment(time_t, time_t)` method to extract events between a specific time interval.

* a place for utilities have been added (`tvheadend::utilities`) which for now contains a file with just a single helper function. The helper provides `erase_if` support for maps, so all the code that loops through things and checks for dirtiness has been compacted into a simple `Utilities::erase_if(...)` call.

* the first two commits are just stuff I found during the way

@ksooo please review if you can, I've only build tested so far. It might help to use `?w=0` when looking at the diff since due to some loops being removed the indentation has changed.

### CODE REVIEW

**Comment by Jalle19** [comment]
The code was originally split into separate commits for each class that was factored out but in the end I had to squash them together to avoid rebase hell.

File: src/tvheadend/entity/AutoRecording.h [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    private:
+      uint32_t m_minDuration; // Minimal duration in seconds (0 = Any).
+      uint32_t m_maxDuration; // Maximal duration in seconds (0 = Any).
+      int32_t  m_approxTime;  // Minutes from midnight (up to 24*60).
+      int64_t  m_startExtra;  // Extra start minutes (pre-time).
+      int64_t  m_stopExtra;   // Extra stop minutes (post-time).
+      uint32_t m_dupDetect;   // duplicate episode detect (record: 0 = all, 1 = different episode number,
+      //                                   2 = different subtitle, 3 = different description,
+      //                                   4 = once per week, 5 = once per day).
```
Comment (ksooo) [inline src/tvheadend/entity/AutoRecording.h]: Indentation broken for above two lines.

File: src/tvheadend/entity/Channel.h [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    typedef std::map<uint32_t, Channel> Channels;
+
+    /**
+     * Represents a channel
+     */
+    class Channel : public Entity
+    {
+    public:
+      Channel() :
+        Entity(),
```
Comment (ksooo) [inline src/tvheadend/entity/Channel.h]: No need to call the base class default ctor directly. Will be done automatically.

File: src/tvheadend/entity/Event.h [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    typedef std::map<uint32_t, Event> Events;
+
+    /**
+     * Represents an event/programme
+     */
+    class Event : public Entity
+    {
+    public:
+      Event() :
+        Entity(),
```
Comment (ksooo) [inline src/tvheadend/entity/Event.h]: No need to call the base class default ctor directly. Will be done automatically. Here and elsewhere...

File: src/tvheadend/entity/Channel.h (hunk 2 above)
Comment (Jalle19) [inline src/tvheadend/entity/Channel.h]: Isn't it necessary when the derived class overrides the constructor?

File: src/tvheadend/entity/AutoRecording.h (hunk 1 above)
Comment (Jalle19) [inline src/tvheadend/entity/AutoRecording.h]: Oh yeah, CLion fucked it up

File: src/tvheadend/entity/RecordingBase.h [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    private:
+      static int GetNextIntId();
+
+      const unsigned int m_iId;
+
+      std::string m_id;         // ID (string!) of dvr[Time|Auto]recEntry.
+      uint32_t m_enabled;    // If [time|auto]rec entry is enabled (activated).
+      uint32_t m_daysOfWeek; // Bitmask - Days of week (0x01 = Monday, 0x40 = Sunday, 0x7f = Whole Week, 0 = Not set).
+      uint32_t m_retention;  // Retention time (in days).
+      uint32_t m_priority;   // Priority (0 = Important, 1 = High, 2 = Normal, 3 = Low, 4 = Unimportant).
```
Comment (ksooo) [inline src/tvheadend/entity/RecordingBase.h]: Indentation of comments (4 lines) broken.

File: src/tvheadend/entity/RecordingBase.h [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+      std::string m_title;      // Title (pattern) for the recording files.
+      std::string m_name;       // Name.
+      std::string m_directory;  // Directory for the recording files.
+      std::string m_owner;      // Owner.
+      std::string m_creator;    // Creator.
+      uint32_t m_channel;    // Channel ID.
```
Comment (ksooo) [inline src/tvheadend/entity/RecordingBase.h]: Indentation of comment broken.

**Comment by ksooo** [comment]
Awesome.

**Comment by Jalle19** [comment]
Had to push some changes:

* forgot to use `utilities::erase_if` in `AutoRecordings` and `TimeRecordings`.
* renamed the `util` directory to `utilities` to match the namespace

**Comment by Jalle19** [comment]
I forgot:

* removed old definitions from `HTSPTypes.h`

**Comment by Jalle19** [comment]
@ksooo rebased and runtime tested extensively

**Comment by Jalle19** [comment]
@ksooo should we get this merged before we start changing things to use setters and getters? Or should I just tack on commits?

**Comment by ksooo** [comment]
I'd opt for adding the getters and setters with another commit and to merge the PR after that.

**Comment by Jalle19** [comment]
Okay, will do.

**Comment by Jalle19** [comment]
@ksooo finally finished the monkey work, can you review at some point? There are a few commits which alter the behavior so you should probably review one commit at a time.

**Comment by ksooo** [comment]
Awesome, despite from the missing m_sid in <code>RecordingBase::operator==</code> I was not able to find anything suspicious.

**Comment by Glenn-1990** [comment]
@Jalle19 Any plant to do the same for SSubscription?

As @ksooo suggested something like this in the other PR and I don't want to conflict with this work.

**Comment by Jalle19** [comment]
@Glenn-1990 yes but not as a part of this PR, so go ahead and change whatever you need in `SSubscription` for now.

**Comment by Jalle19** [comment]
@ksooo do you mind taking this for a test drive? I just spent over an hour rebasing it and managed to find a bug in the initial refactoring while I was at it (see last commit).

**Comment by ksooo** [comment]
@Jalle19 please, give me some time for this and ... please rebase.

**Comment by Jalle19** [comment]
I'm thinking we should try to merge this before I add any more commits to it, otherwise it will be even harder to rebase it.

**Comment by ksooo** [comment]
Makes sense. But during the next days I will hardly find the time to runtime test this.

But still ... rebase please, then I can pick this up for testing.

**Comment by Jalle19** [comment]
Will rebase later today. There's some bug where timers that are part of autorec entries have the UUID as their title, but I don't see how that could be because of the new code. Have you seen that with the current master too?

**Comment by ksooo** [comment]
> part of autorec entries have the UUID as their title

Which tvheadend version are you using? Maybe something has changed (broke) in recent tvh? https://github.com/tvheadend/tvheadend/commit/220833201fdd56dc25bfbdedbf7633ceb19b63b6 changed a lot, also in the area of autorec titles / names.

**Comment by Jalle19** [comment]
@ksooo 4.0.4 I think

**Comment by Jalle19** [comment]
@ksooo let's try to get this shoved in during devcon. I have some things I'd like to work on and I'd rather not have to rebase this more than necessary.

**Comment by ksooo** [comment]
:+1:

**Comment by Jalle19** [comment]
@ksooo it's rebased now and I found the issue with the auto recording title (turns out it was a rebase error). Can you please test this since there is a slight possibility I missed something that was added recently while rebasing?

**Comment by Jalle19** [comment]
Squashed an update to CMakeList.txt.

**Comment by ksooo** [comment]
Did a quick runtime test. Couldn't find any regression. Time to shove this in, I'd say ... and worth a new version number and changelog maybe.

**Pull request merged**